packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp

```diff
@@ -113,1133 +695,1184 @@ setupArraysForNodeRule(const Teuchos::RCP<const panzer::IntegrationRule>& ir)
 
   Intrepid2::CellTools<PHX::Device::execution_space> cell_tools;
+  MDFieldArrayFactory af(prefix,true);
 
-  const shards::CellTopology & cell_topology = *(int_rule->topology);
-  const panzer::IntegrationRule & ir = *int_rule;
+  const int num_space_dim = int_rule->topology->getDimension();
+  const int num_ip = int_rule->num_points;
+  const int num_nodes = int_rule->topology->getNodeCount();
 
-  const int num_cells = in_num_cells < 0 ? in_node_coordinates.extent(0) : in_num_cells;
+  auto aux = af.template buildStaticArray<Scalar,Cell,IP>("weighted_measure",num_cells_, num_ip);
 
-  // Copy over coordinates
-  {
-    const int num_nodes = in_node_coordinates.extent(1);
-    const int num_dims = in_node_coordinates.extent(2);
-    auto node_coordinates_k = node_coordinates.get_view();
-    auto in_node_coordinates_k = in_node_coordinates.get_view();
-
-    Kokkos::MDRangePolicy<PHX::Device::execution_space,Kokkos::Rank<3>> policy({0,0,0},{num_cells,num_nodes,num_dims});
-    Kokkos::parallel_for("node_coordinates",policy,KOKKOS_LAMBDA (const int cell, const int node, const int dim) {
-      node_coordinates_k(cell,node,dim) = in_node_coordinates_k(cell,node,dim);
-    });
-    PHX::Device::execution_space().fence();
-  }
+  if(int_rule->cv_type != "none"){
 
-  // NOTE: We are assuming that each face can have a different number of points.
-  // Not sure if this is necessary, but it requires a lot of additional allocations
+    TEUCHOS_TEST_FOR_EXCEPT_MSG(int_rule->cv_type == "side",
+                                "IntegrationValues2::getWeightedMeasure : Weighted measure not available for side control volume integrators. Use getWeightedNormals instead.");
 
-  const int cell_dim = cell_topology.getDimension();
-  const int subcell_dim = cell_topology.getDimension()-1;
-  const int num_subcells = cell_topology.getSubcellCount(subcell_dim);
+    // CV integration uses a single call to map from physical space to the weighted measure - I assume this is slower than what we do with non-cv integration methods
 
-  Intrepid2::DefaultCubatureFactory cubature_factory;
+    auto s_cub_points = af.template buildStaticArray<Scalar, Cell, IP, Dim>("cub_points",num_evaluate_cells_,num_ip,num_space_dim);
 
-  // We get to build up our cubature one face at a time
-  {
-    int point_offset=0;
-    for(int subcell_index=0; subcell_index<num_subcells; ++subcell_index) {
+    const auto cell_range = std::make_pair(0,num_evaluate_cells_);
+    auto s_node_coord = Kokkos::subdynrankview(getNodeCoordinates().get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL());
+    auto s_weighted_measure = Kokkos::subdynrankview(aux.get_view(),           cell_range,Kokkos::ALL());
+
+    // FIXME: getCubature call requires a non-const node coordinates array
+    auto nonconst_node_coord = af.template buildStaticArray<Scalar, Cell, NODE, Dim>("scratch",num_evaluate_cells_,num_nodes,num_space_dim);
+    Kokkos::deep_copy(nonconst_node_coord.get_view(), s_node_coord);
```

Comment:
Another place to use data() to create an unmanaged view.

Reply:
I believe I've hit all of them.